ducktape/tests/runner.py

```diff
@@ -61,46 +68,98 @@ def __init__(self, cluster, session_context, session_logger, tests, port=Console
         self.receiver = Receiver(port)
 
         self.session_context = session_context
+        self.max_parallel = session_context.max_parallel
         self.results = TestResults(self.session_context)
 
-        self.proc_list = []
-        self.staged_tests = tests
+        self.main_process_pid = os.getpid()
+        self.scheduler = TestScheduler(
+            [TestExpectedNodes(test_context=t, expected_nodes=self.expected_num_nodes(t)) for t in tests],
+            self.cluster)
         self._test_context = {t.test_id: t for t in tests}
         self._test_cluster = {}  # Track subcluster assigned to a particular test_id
+        self._client_procs = {}  # track client processes running tests
         self.active_tests = {}
         self.finished_tests = {}
 
+    def _propagate_sigterm(self, signum, frame):
+        """Handler SIGTERM and SIGINT by propagating SIGTERM to all client processes.
+
+        Note that multiprocessing processes are in the same process group as the main process, so Ctrl-C will
+        result in SIGINT being propagated to all client processes automatically. This may result in multiple SIGTERM
+        signals getting sent to client processes in quick succession.
+
+        However, it is possible that the main process (and not the process group) receives a SIGINT or SIGTERM
+        directly. Propagating SIGTERM to client processes is necessary in this case.
+        """
+
+        if os.getpid() != self.main_process_pid:
+            # since we're using the multiprocessing module to create client processes,
+            # this signal handler is also attached client processes, so we only want to propagate TERM signals
+            # if this process *is* the main runner server process
+            return
+
+        self.stop_testing = True
+        for p in self._client_procs.values():
+
+            # this handler should be a noop if we're in a client process, so it's an error if the current pid
+            # is in self._client_procs
+            assert p.pid != os.getpid(), "Signal handler should not reach this point in a client subprocess."
+            if p.is_alive():
+                os.kill(p.pid, signal.SIGTERM)
+
     def who_am_i(self):
         """Human-readable name helpful for logging."""
         return self.__class__.__name__
 
+    @property
+    def _ready_to_trigger_more_tests(self):
+        """Should we pull another test from the scheduler?"""
+        return not self.stop_testing and \
+            len(self.active_tests) < self.max_parallel and \
+            len(self.scheduler) > 0 and \
```

Comment:
Yeah, my point was just that since you only ever use `peek()` having already checked `len(self.scheduler) > 0` (as here), the exception is never actually used and you could just remove that code. In any case, either approach works, returning `None` in both cases would just require moving the check of `len(self.scheduler)` to the `while` loop condition.